wgengine/tsdns/tsdns.go

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package tsdns provides a Resolver struct capable of resolving
+// domains on a Tailscale network.
+package tsdns
+
+import (
+	"encoding/binary"
+	"errors"
+	"sync"
+
+	"github.com/tailscale/wireguard-go/device"
+	dns "golang.org/x/net/dns/dnsmessage"
+	"inet.af/netaddr"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/packet"
+)
+
+// MaxResponseSize is the maximal size of a Tailscale DNS response.
+const MaxResponseSize = 512
+
+const (
+	// ipOffset is the space before the IP header. It is reserved for wireguard-go.
+	ipOffset = device.MessageTransportHeaderSize
+	// dnsDataOffset is the space before DNS data. It includes the IP and UDP headers.
+	dnsDataOffset = ipOffset + packet.UDPDataOffset
+
+	// The response buffer must have space for all the headers and the response body.
+	bufferSize = dnsDataOffset + MaxResponseSize
+)
+
+var (
+	errMapNotSet      = errors.New("domain map not set")
+	errNoSuchDomain   = errors.New("domain does not exist")
+	errNotImplemented = errors.New("query type not implemented")
+	errNotOurName     = errors.New("not an *.ipn.dev domain")
+	errNotQuery       = errors.New("not a DNS query")
+)
+
+var (
+	// The default IP for a new resolver.
+	DefaultIP = packet.IP(binary.BigEndian.Uint32([]byte{100, 100, 100, 100}))
+	// The default port for a new resolver.
+	DefaultPort = uint16(53)
+)
+
+// DNSMap is all the data Resolver needs to resolve DNS queries.
+type DNSMap struct {
```

Comment:
This is already in a package named "tsdns", so the leading "DNS" is redundant/stuttery. (https://golang.org/doc/effective_go.html#package-names)

Other users will always use the `tsdns.` prefix, so `tsdns.Map` looks nicer & shorter.